internal/thelma/utils/pwgen/pwgen.go

```diff
@@ -0,0 +1,92 @@
+package pwgen
+
+import (
+	"github.com/broadinstitute/thelma/internal/thelma/app/logging"
+	"github.com/broadinstitute/thelma/internal/thelma/app/seed"
+	"math/rand"
+)
+
+const defaultLength = 24
+const absoluteMinimumLength = 8
+
+func init() {
+	seed.Rand()
+}
+
+var lower = []rune(`abcdefghijklmnopqrstuvwxyz`)
+var upper = []rune(`ABCDEFGHIJKLMNOPQRSTUVWXYZ`)
+var num = []rune(`1234567890`)
+var special = []rune(`!@#$%^&*()_-+={}[]/\<>.,;?:|'"`)
+
+var all = func() []rune {
+	var a []rune
+	a = append(a, lower...)
+	a = append(a, upper...)
+	a = append(a, num...)
+	a = append(a, special...)
+	return a
+}()
+
+type Pwgen struct {
+	MinLength  int
+	MinLower   int
+	MinUpper   int
+	MinNum     int
+	MinSpecial int
+}
+
+func (p Pwgen) Generate() string {
+	l := p.MinLength
+	if l <= 0 {
+		l = defaultLength
+	}
+	if l < absoluteMinimumLength {
+		l = absoluteMinimumLength
+	}
+
+	minChars := p.MinLower + p.MinUpper + p.MinNum + p.MinSpecial
+	if l < minChars {
+		l = minChars
+	}
+
+	buf := make([]rune, l)
+
+	i := 0
+
+	n := p.MinLower
+	for ; i < n; i++ {
+		buf[i] = pick(lower)
+	}
+
+	n += p.MinUpper
+	for ; i < n; i++ {
+		buf[i] = pick(upper)
+	}
+
+	n += p.MinNum
+	for ; i < n; i++ {
+		buf[i] = pick(num)
+	}
+
+	n += p.MinSpecial
+	for ; i < n; i++ {
+		buf[i] = pick(special)
+	}
+
+	n = l
+	for ; i < n; i++ {
+		buf[i] = pick(all)
+	}
+
+	rand.Shuffle(n, func(a, b int) {
+		buf[a], buf[b] = buf[b], buf[a]
+	})
+
+	password := string(buf)
+	logging.MaskSecret(password)
+	return password
+}
+
+func pick(from []rune) rune {
+	return from[rand.Intn(len(from))]
```

Comment:
this is calling math/rand instead of crypto/rand; suppose maybe it's not a big deal. crypto/rand doesn't have a shuffle unfortunately. suppose another approach could be just to b64 a bunch of random bytes out of crypto/rand but i guess the result would end up being longer to achieve good complexity

Reply:
Thanks - I admit I was being lazy because math/rand is more user-friendly :) But I see a few blog posts about using crypto rand as a source for math rand, will revise quickly.